DEVELOPERS.md

```diff
@@ -0,0 +1,122 @@
+
+### Metadata Update Replication
+This will update the metadata only for a table that was previously fully replicated.
+
+No data will be copied but any metadata from the source will be copied and table/partition locations will keep pointing to previously replicated data.
+
+Example use case: Update the metadata of a Hive Table (for instance to change the Serde used) without having the overhead of re-replicating all the data.
+
+## Copiers
+The copiers are the classes which do the actual copying of the data. 
+
+There is a `CopierFactoryManager` which determines which type of copier will be used. The `DefaultCopierFactoryManager` is an implementation of this, and has a list of `CopierFactories` auto-wired into it. Spring will find all beans which are implementations of the `CopierFactory` and pass these into the constructor for the `DefaultCopierFactoryManager`. 
+
+There is an optional copier option available to set which `CopierFactory` to use, if this value is set this copier factory class will be used. If this value is not set the `DefaultCopierFactoryManager` will check all `CopierFactories` in the list and return the first which supports replication between the SourceLocation and ReplicaLocation provided. 
+
+There is an order of precedence, which means the `CopierFactories` will be checked in the following order to see if the replication is supported:
+* `S3S3Copier`,
+* `S3MapreduceCpCopier`, 
+* and then falls down to `DistCpCopier` if the above factories don't support the replication.
+
+The copiers which use S3 will create clients that allow access and give permissions to perform actions on S3 buckets. In some cases an IAM role is needed, if data is being transferred across S3 accounts. 
+
+### Types of copier
+**S3S3Copier**
+
+*Replication: S3 → S3* 
+
+This copier uses two `AwsS3Clients` - a source client and a replica client. There is an `AwsS3ClientFactory` which will create clients with the necessary permissions to perform actions on S3 buckets
+
+One of these client factories is `JceksAmazonS3ClientFactory`, which creates a client with the necessary credentials required. It does this using a credential provider chain, which will create (as the name states) a chain of credential providers which will be tried in order, until one is successful. One of the credentials in this chain is the `AssumeRoleCredentialProvider` which uses a role provided in the copier options to be able to replicate across S3 accounts.
+
+The replication is handled by a `TransferManager` which uses the target S3 client and the `S3S3CopierOptions`. The `TransferManager` will be given the the source client to replicate from. 
+
+The `S3S3CopierOptions` will take the `CopierOptions` provided and change them into more specific s3 options. For example it will have the options `s3-server-side-encryption` and `assume-role`, which are specific to S3 clients and wont be used by the other copiers. 
```

Comment:
```suggestion
The `S3S3CopierOptions` will take the `CopierOptions` provided and change them into more specific s3 options. For example it will have the options `s3-server-side-encryption` and `assume-role`, which are specific to S3 clients and won't be used by the other copiers. 
```